pkg/archive/archive.go

```diff
@@ -148,12 +149,18 @@ func DecompressStream(archive io.Reader) (io.ReadCloser, error) {
 		readBufWrapper := p.NewReadCloserWrapper(buf, bz2Reader)
 		return readBufWrapper, nil
 	case Xz:
-		xzReader, err := xzDecompress(buf)
+		xzReader, chdone, err := xzDecompress(buf)
 		if err != nil {
 			return nil, err
 		}
 		readBufWrapper := p.NewReadCloserWrapper(buf, xzReader)
-		return readBufWrapper, nil
+		return ioutils.NewReadCloserWrapper(readBufWrapper, func() error {
+			go func() {
```

Comment:
first, have you reproduced the issue #16924 the `[panic: runtime error: slice bounds out of range]` crash? if docker-untar something failed, that crash is supposed to happen;

in this fix, I do wait channel and Close in a separate goroutine, that fixed the crash; but if it's not in a separate goroutine, code like this ends up deadlock, docker build on `ADD broken.tar.xz .." never returns

```go
return ioutils.NewReadCloserWrapper(readBufWrapper, func() error {
    <-chdone
    return readBufWrapper.Close()
}), nil
```